pulser/parametrized/variable.py

```diff
@@ -0,0 +1,121 @@
+# Copyright 2020 Pulser Development Team
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass
+from typing import Union
+
+import numpy as np
+
+from pulser.parametrized import Parametrized
+from pulser.parametrized.paramobj import OpSupport
+
+
+@dataclass(frozen=True, eq=False)
+class Variable(Parametrized, OpSupport):
+    """A variable for parametrized sequence building.
+
+    Args:
+        name (str): Unique name for the variable.
+        dtype (type): Type of the variable's content. Supports `float`, `int`
+            and `float`.
+        size (int=1): The number of values stored. Defaults to a single value.
+    """
+
+    name: str
+    dtype: type
+    size: int = 1
+
+    def __post_init__(self):
+        if not isinstance(self.name, str):
+            raise TypeError("Variable's 'name' has to be of type 'str'.")
+        if self.dtype not in [int, float, str]:
+            raise TypeError(f"Invalid data type '{self.dtype}' for Variable.")
+        if not isinstance(self.size, int):
+            raise TypeError("Given variable 'size' is not of type 'int'.")
+        elif self.size < 1:
+            raise ValueError("Variables must be of size 1 or larger.")
+        self.__dict__["_count"] = -1      # Counts the updates
+        self._clear()
+
+    @property
+    def variables(self):
+        return {self.name: self}
+
+    def _clear(self):
+        self.__dict__["value"] = None
+        self.__dict__["_count"] += 1
+
+    def _assign(self, value):
+        if self.dtype == str:
+            if not (isinstance(value, str) if self.size == 1
+                    else all(isinstance(s, str) for s in value)):
+                raise TypeError(f"Provided values for variable '{self.name}' "
+                                "must be of type 'str'.")
+
+        val = np.array(value, dtype=self.dtype)
+        if val.size != self.size:
+            raise ValueError(f"Can't assign array of size {val.size} to "
+                             + f"variable of size {self.size}.")
+
+        self.__dict__["value"] = self.dtype(val) if self.size == 1 else val
```

Comment:
Yes, because I made `Variable` a `dataclass` with `frozen=True`. This makes it behave like an immutable object, so the user can't change its attributes (eg. assigns a value to the variable directly with `var.value=...`). 
It does this by blocking the `__setattr__()` and `__delattr__()`, so you have to assign values in this way. I know it's not pretty, but it works.

Reply:
Ah got it !